### PR TITLE
Add resource pack, texture pack and shader pack downloaders

### LINKED ISSUE #153: Resource / texture pack downloader
### Role

I like resource packs

### Suggestion

Be able to download resource packs from modrinth and curseforge

### Benefit

I don't have to open a web browser, or have to check for updates

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### LINKED ISSUE #725: Shader integration with Modrinth
### Role

I play Minecraft with shaders.

### Suggestion

I want the Shaders panel to function like the Mods panel, and sync with Modrinth.

### Benefit

It makes Shader updates and acquisitions easier.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

*This may be considered a dupe of #153 but that is regarding resource packs.*

Modrinth now has support for Shaders, so implementation with Modrinth would be convenient, as well as making sure you have the correct shader engine (Iris, OptiFine, Canvas, etc) installed in order to use them.

### PR DESCRIPTION
**Pull request by flowln**
Closes #153
Closes #725

TODO:
- [x] Resource packs (is 'RPs' a good smaller name? couldn't figure out a better one to put in the button)
  - [x] MR
  - [x] CF
- [x] Shader packs
  - [x] MR
  - No CF since we can't filter shader packs in the API
- [x] Texture packs (maybe just extending the RP ones can work out, haven't tested yet tho)
  - [x] MR (Filters the versions > 1.6 out in the API)
  - [x] CF (Filters to only show the version of the instance in the API, since we can't do multiple versions in the API, and it gets too clogged with all resource packs)

no Worlds and Data packs yet since we need to refactor them into the Resource abstraction before.

Shaders are a bit hard since they often require other mods to work properly, but hard-coding those doesn't seem like a good approach. I guess this could be addressed by adding loader filters, like the one we have for mod's MC versions, but we need to refactor that as well in order to not create a bunch of spaghetti code ig

note to self: remember to add the SPDX license thingies before opening the PR for review

### CODE REVIEW

**Comment by RaptaG** [comment]
Could a request be made in the CF team to categorize shaders in the API?

**Comment by flowln** [comment]
> Could a request be made in the CF team to categorize shaders in the API?

I doubt it would be simple to them to do so, considering they already have shaders in there, just not in a 'Shaders-only' category, so they would need to re-categorize every one of those. I guess someone could try their luck and ask them to do so in their suggestions platform, but I doubt it would go anywhere in the foreseeable future.

**Comment by TheKodeToad** [comment]
Yeah... even on the website there is no shaders category. Shaders have no special file extension so they can't be filtered like that.

**Comment by DioEgizio** [comment]
> > Could a request be made in the CF team to categorize shaders in the API?
> 
> I doubt it would be simple to them to do so, considering they already have shaders in there, just not in a 'Shaders-only' category, so they would need to re-categorize every one of those. I guess someone could try their luck and ask them to do so in their suggestions platform, but I doubt it would go anywhere in the foreseeable future.

Also rn you just can't download shaders at all, there's no downloads for them

**Comment by Ryex** [comment]
> > > Could a request be made in the CF team to categorize shaders in the API?
> > 
> > 
> > I doubt it would be simple to them to do so, considering they already have shaders in there, just not in a 'Shaders-only' category, so they would need to re-categorize every one of those. I guess someone could try their luck and ask them to do so in their suggestions platform, but I doubt it would go anywhere in the foreseeable future.
> 
> Also rn you just can't download shaders at all, there's no downloads for them

can we get their listings at all? we may be able to  at lest point people to their downloads page.

**Comment by flowln** [comment]
> can we get their listings at all? we may be able to at lest point people to their downloads page.

not in a clean way. AFAIK they don't have a 'Shader' project type in the API that we can filter for, it's mixed with other types of stuff in the ['Customization' category](https://www.curseforge.com/minecraft/customization), so if we were to list the projects, there would also be a lot of non-shader projects in there too.

**Comment by Frontear** [comment]
Adding the tag "shader" into the customization search filter does help, but I doubt that's a maintainable solution

**Comment by Ryex** [comment]
... I did some API tests

find the catagory:
```json
https://api.curseforge.com/v1/categories?gameId=432
    {
      "id": 4547,
      "gameId": 432,
      "name": "Configuration",
      "slug": "configuration",
      "url": "https://www.curseforge.com/minecraft/customization/configuration",
      "iconUrl": "https://media.forgecdn.net/avatars/25/841/635781940944540686.png",
      "dateModified": "2015-09-18T22:28:14.453Z",
      "classId": 4546,
      "parentCategoryId": 4546,
      "displayIndex": 0
    },

```


``` json
https https://api.curseforge.com/v1/mods/search?gameId=432&classID=4546
{
    "data": [],
    "pagination": {
        "index": 0,
        "pageSize": 50,
        "resultCount": 0,
        "totalCount": 0
    }
}
```

you strait up *can't* pull customization items form the API....

**Comment by DioEgizio** [comment]
> ... I did some API tests
> 
> find the catagory:
> 
> ```json
> https://api.curseforge.com/v1/categories?gameId=432
>     {
>       "id": 4547,
>       "gameId": 432,
>       "name": "Configuration",
>       "slug": "configuration",
>       "url": "https://www.curseforge.com/minecraft/customization/configuration",
>       "iconUrl": "https://media.forgecdn.net/avatars/25/841/635781940944540686.png",
>       "dateModified": "2015-09-18T22:28:14.453Z",
>       "classId": 4546,
>       "parentCategoryId": 4546,
>       "displayIndex": 0
>     },
> ```
> 
> ```json
> https https://api.curseforge.com/v1/mods/search?gameId=432&classID=4546
> {
>     "data": [],
>     "pagination": {
>         "index": 0,
>         "pageSize": 50,
>         "resultCount": 0,
>         "totalCount": 0
>     }
> }
> ```
> 
> you strait up _can't_ pull customization items form the API....

That's what I already said... No point on implementing shader pack support from curseforge with some weird workarounds IMO

**Comment by Madis0** [comment]
In what way do texture packs need a lot of special handling in this context? To me it just seems
```
if(mcVersion >= a1.1.2 && mcVersion <= 1.6)
   downloadTo("texturepacks");
else if(mcVersion >= 1.6.1)
    downloadTo("resourcepacks");
```

**Comment by flowln** [comment]
> In what way do texture packs need a lot of special handling in this context? To me it just seems
> 
> ```
> if(mcVersion >= a1.1.2 && mcVersion <= 1.6)
>    downloadTo("texturepacks");
> else if(mcVersion >= 1.6.1)
>     downloadTo("resourcepacks");
> ```

Ideally we would only show texture packs in its downloader, but then it needs to be filtered out in the API search from the resource packs. Besides, whether an instance uses texture packs or resource packs depends on a `trait` attribute in meta, so if we were to simply do a version comparison, it would duplicate that logic on both meta and in the launcher, which is not ideal. So, what we need to do is to get all the MC versions in meta, and filter them by their trait, and include only those versions in the API search.

**Comment by DioEgizio** [comment]
> > In what way do texture packs need a lot of special handling in this context? To me it just seems
> > ```
> > if(mcVersion >= a1.1.2 && mcVersion <= 1.6)
> >    downloadTo("texturepacks");
> > else if(mcVersion >= 1.6.1)
> >     downloadTo("resourcepacks");
> > ```
> 
> Ideally we would only show texture packs in its downloader, but then it needs to be filtered out in the API search from the resource packs. Besides, whether an instance uses texture packs or resource packs depends on a `trait` attribute in meta, so if we were to simply do a version comparison, it would duplicate that logic on both meta and in the launcher, which is not ideal. So, what we need to do is to get all the MC versions in meta, and filter them by their trait, and include only those versions in the API search.

We might wanna disable all version filter for CF tpacks then as we can't search for more than a version at a time on CF api

**Comment by flowln** [comment]
> We might wanna disable all version filter for CF tpacks then as we can't search for more than a version at a time on CF api

yea, we might wanna do a similar thing to the mod version filter we have rn, so that it at least shows the knowingly compatible ones right from the start, and move the "include-all-packs-even-if-they-are-resource-packs" option to the 'All versions' filter. Not gonna do filters in this PR though.

**Comment by flowln** [comment]
Upon closer inspection of the data available to us, I noticed we don't actually have the `traits` fields in the index file, so we would have to download the meta json for every version in order to look at that and filter the versions that use texture packs. So, I just used a simple version comparison in the end, and added a FIXME because it duplicates the logic in meta. We could probably aggregate such info in an index file in meta, but this works fine for now :pineapple:

The CF version filtering is a bit ad-hoc, but I think it's better than showing every resource pack, since most of those don't have pre-1.6.1 versions, and it just gets annoying to find the ones that do.

**Comment by flowln** [comment]
Blocked until #751 gets merged since this relies a lot on the version parsing for the texture pack version thingies

**Comment by Scrumplex** [comment]
> Blocked until #751 gets merged since this relies a lot on the version parsing for the texture pack version thingies

I ~~scrumped~~ merged it!

**Comment by TheKodeToad** [comment]
[Scrumplex](https://github.com/Scrumplex) scrumped commit [a47bf72](https://github.com/PrismLauncher/PrismLauncher/commit/a47bf72b07c73a28982dcdef6961dce5bf704702) into PrismLauncher:develop

(on topic)

**Comment by RaptaG** [comment]
There will be no resource pack / shader pack updater I assume, right?

**Comment by flowln** [comment]
> There will be no resource pack / shader pack updater I assume, right?

not in this PR, maybe it can be added at some later point in time

**Review by DioEgizio — APPROVED** [review]

File: launcher/ui/pages/instance/ResourcePackPage.cpp
Comment (Scrumplex) [inline launcher/ui/pages/instance/ResourcePackPage.cpp]:
```suggestion
    ui->actionDownloadItem->setText(tr("Download packs"));
    ui->actionDownloadItem->setToolTip(tr("Download resource pacs from online platforms"));
```

File: launcher/ui/pages/instance/TexturePackPage.cpp
Comment (Scrumplex) [inline launcher/ui/pages/instance/TexturePackPage.cpp]:
```suggestion
    ui->actionDownloadItem->setText(tr("Download packs"));
    ui->actionDownloadItem->setToolTip(tr("Download texture packs from online platforms"));
```

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Great work!

**Comment by RaptaG** [comment]
O.O

**Pull request merged**
